### PR TITLE
Add support for SSLKEYLOGFILE

### PR DESCRIPTION
**Pull request by lukebakken**
See #132

### CODE REVIEW

File: lib/finch/http1/conn.ex
Comment (sneako) [inline lib/finch/http1/conn.ex:212]: Looks like we could just use `HTTP.get_socket/1` for both HTTP1 & HTTP2. Then the implementation would be exactly the same for both

File: lib/finch/util.ex
Comment (sneako) [inline lib/finch/util.ex:1]: Rather than creating a generic Util module, I would prefer a dedicated module, maybe something like `Finch.SslKeyLog`?

File: lib/finch/util.ex
Comment (sneako) [inline lib/finch/util.ex:5]: I don't think we should read the system env on every single https request. Couldn't we add a new conn_opt instead?

File: lib/finch/util.ex
Comment (sneako) [inline lib/finch/util.ex:23]: How does this behave in a highly concurrent environment, ie: if there are many processes concurrently trying to append to the same file?

**Review by sneako — CHANGES_REQUESTED** [review]
Hi and thanks for the PR! I definitely agree that this feature would be very useful as I have also had the need to decode packets in wireshark for https requests.

There are a few things I think we should change about the implementation, the main thing being the use of the system ENV to toggle this behaviour. Also, I would never be able to use this as is in production because there would be way too much data. Maybe we should also consider adding a sample rate parameter?

File: lib/finch/util.ex
Comment (lukebakken) [reply lib/finch/util.ex:23]: Good point, I will fix that. Lines can be interleaved but obviously writing a line from `:keylog` has to be atomic.

File: lib/finch/util.ex
Comment (lukebakken) [reply lib/finch/util.ex:5]: A configuration setting makes more sense!

**Comment by lukebakken** [comment]
@sneako thanks for taking a look.

> There are a few things I think we should change about the implementation, the main thing being the use of the system ENV to toggle this behaviour. 

One reason to use `SSLKEYLOGFILE` is that it is "standard" in that browsers and the NSS library recognize it - https://developer.mozilla.org/en-US/docs/Mozilla/Projects/NSS/Key_Log_Format. Of course, a configuration option passed to Finch would do as well. I discovered `SSLKEYLOGFILE` when I was doing some Wireshark tracing of browser-based HTTP2 sessions. I suspect that users will discover this feature in a similar manner so the code or documentation in Finch should reflect that. At any rate, I will switch to a configuration setting ... how about `ssl_key_log_file`? I will make sure the docs are clear about it.

**Comment by wojtekmach** [comment]
I think it's fine to use the `SSLKEYLOGFILE` env var but it's important it is only read once, when pool starts, and not on every request. But an option is even better.

**Comment by lukebakken** [comment]
@sneako ready for a re-review. Thank you!

File: test/finch/http1/integration_test.exs
Comment (wojtekmach) [inline test/finch/http1/integration_test.exs:43]: no need for asserts here imo:

```suggestion
      tmp_dir = System.tmp_dir()
      log_file = Path.join(tmp_dir, "ssl-key-file.log")
      :ok = System.put_env("SSLKEYLOGFILE", log_file)
```

File: test/finch/http1/integration_test.exs
Comment (wojtekmach) [inline test/finch/http1/integration_test.exs:65]:
```suggestion
        assert File.stat!(log_file).size > 0
```

File: test/finch/http1/integration_test.exs
Comment (wojtekmach) [inline test/finch/http1/integration_test.exs:67]:
```suggestion
        File.rm!(log_file)
```

**Comment by sneako** [comment]
Thanks @wojtekmach, I agree with all of you points. And sorry everyone, it has been a busy week for me. One thing that hasn't been addressed yet was my comment re: sample_rate. If I tried to use this in production, I'm sure my node would be overloaded, so I think an optional sample_rate option would be one way around that. What do you all think?

**Comment by wojtekmach** [comment]
@sneako oh, forgot to mention, I don't think it's going to be necessarily problematic in prod performance-wise, my understanding is we only log the keys on connect once per connection, not on every request.

**Comment by lukebakken** [comment]
Thanks everyone. I'll try to get these changes soon, but please note that you do have the ability to make changes to my branch and push directly to my fork if you'd like.

Logging secrets is not per-request, but per connection.

**Comment by sneako** [comment]
Ah perfect, disregard in that case 😄

**Comment by lukebakken** [comment]
@wojtekmach @sneako all set for re-review. For some reason GitHub isn't allowing me to re-request a review 🤷

File: lib/finch.ex
Comment (sneako) [inline lib/finch.ex:8]: This alias is unused
```suggestion
  alias Finch.{PoolManager, Request, Response}
```

File: lib/finch.ex
Comment (sneako) [inline lib/finch.ex:52]: Could you please use this formatting?
```
      doc: """
      These options are passed to `Mint.HTTP.connect/4` whenever a new connection is established. \
      `:mode` is not configurable as Finch must control this setting. Typically these options are \
      used to configure proxying, https settings, or connect timeouts. The `ssl_key_log_file` \
      connection option specifies a file into which all TLS secrets are logged for later use in a \
      tool like Wireshark to decrypt https sessions. Setting the `SSLKEYLOGFILE` environment \
      variable also enables this feature.
      """,
```

**Comment by sneako** [comment]
Just two small comments, but otherwise I think this is good to go! We need to wait for https://github.com/elixir-mint/mint/pull/314 to be merged & released first though, right?

**Comment by lukebakken** [comment]
> We need to wait for elixir-mint/mint#314 to be merged & released first though, right?

Actually, `get_socket/1` is available in `mint` right now, it's just not officially "public" or with tests like I added in that PR.

Thanks for the review, I'll get the changes today.

**Comment by lukebakken** [comment]
@sneako I just noticed the previous workflow failed due to TLS cypher issues. Could you re-trigger the workflow? It appears to be triggered by the particular system/OTP combo used in GH Actions. Thanks.

**Comment by lukebakken** [comment]
Argh, even with 35d8776 I still see `:insufficient_crypto_support` in one of the CI workflow runs. Trying something else.

**Comment by lukebakken** [comment]
@wojtekmach @sneako this is ready for re-review. Thanks in advance!

**Comment by sneako** [comment]
Hey @lukebakken ! So I checked out the PR locally and tried to get the TLS v1.3 tests to run on OTP 24, but ran in to some dependency issues with some of our test deps. So instead I tried out the feature (on OTP 24), both via the conn_opts and via the env var, and it worked well! I'll figure out the dependency issues as they get updated to support OTP 24, but I don't think we need to wait for that.

I felt that the docs you added warranted their own section, so I made that change and pushed. Please let me know what you think, and if it looks good, I will merge this and release Finch `0.7.0` 😄

**Review by sneako — APPROVED** [review]

**Comment by sneako** [comment]
Ah perhaps I misunderstood the note here: https://github.com/keathley/finch/pull/133/commits/811f16b75df28ee79e583ab9b3738600219b1405 ? Shouldn’t these failing tests be skipped when TLS v1.3 is not supported anyway?

**Comment by lukebakken** [comment]
@sneako removing those cases from the tests is great. I need to keep bugging the `mint` team for version 1.3

**Comment by lukebakken** [comment]
OK great, I'll fix the conflicts.

**Comment by lukebakken** [comment]
Still failing 😡 ... the main difference I see between GH actions in this project vs `elixir-mint/mint` is that Erlang/OTP is built on `ubuntu-14.04`, while in `mint` it is `ubuntu-18.04`. Continuing to investigate.

**Comment by lukebakken** [comment]
@wojtekmach I was going to switch to using `erlef/setup-elixir@v1`, the same step that `mint` uses. What do you think?

**Comment by wojtekmach** [comment]
yeah, that was the culprit. I was going to add some minor changes on top but they're optional, I'll send a separate PR after this lands.

**Comment by sneako** [comment]
Sounds good @wojtekmach I'll hold off on publishing the release in that case. Thanks for the PR @lukebakken !

File: .github/workflows/elixir.yml
Comment (sneako) [reply .github/workflows/elixir.yml:8]: 👍 I don't think we need both

**Pull request merged**
